homeassistant/components/tuya/binary_sensor.py

```diff
@@ -348,6 +353,48 @@ async def async_setup_entry(
     """Set up Tuya binary sensor dynamically through Tuya discovery."""
     hass_data = entry.runtime_data
 
+    def _get_fault_labels(device: CustomerDevice) -> list[str]:
+        """Return fault labels for the device."""
+        if DPCode.FAULT not in device.status_range:
+            return []
+
+        fault_type = DPType(device.status_range[DPCode.FAULT].type)
+        if fault_type != DPType.BITMAP:
+            return []
+
+        fault_values = json.loads(device.status_range[DPCode.FAULT].values)
+        if "label" not in fault_values:
+            return []
+
+        fault_labels = fault_values["label"]
+        if not isinstance(fault_labels, list) or not all(
+            isinstance(label, str) for label in fault_labels
+        ):
+            return []
+
+        return fault_labels
+
+    def _get_fault_sensors(device: CustomerDevice) -> list[TuyaBinarySensorEntity]:
+        """Return fault sensors for the device."""
+        fault_labels = _get_fault_labels(device)
+        return [
+            TuyaBinarySensorEntity(
+                device,
+                hass_data.manager,
+                TuyaBinarySensorEntityDescription(
+                    name=fault_label,
+                    translation_key=fault_label,
+                    key=f"{DPCode.FAULT}_{fault_label}",
+                    dpcode=DPCode.FAULT,
+                    fault_key=fault_label,
+                    fault_keys=fault_labels,
+                    device_class=BinarySensorDeviceClass.PROBLEM,
+                    entity_category=EntityCategory.DIAGNOSTIC,
+                ),
+            )
+            for fault_label in fault_labels
+        ]
```

Comment:
Let's just inline this logic in the entry setup.

Reply:
I've inlined it in `async_discover_device` now 👍